containerd integration: `docker commit`

laurazard opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Implement `docker commit` with containerd store

**- How I did it**

Upstream (with some adjustments):
- https://github.com/rumpl/moby/pull/37
- https://github.com/rumpl/moby/pull/76
- https://github.com/rumpl/moby/pull/110

**- How to verify it**

(run a container and create a file)
- `docker run -it --name hello alpine`
- `/ # echo world > world.txt && exit`
- `docker commit hello hello/world`

(verify file was written to new image)
- `docker run hello/world cat world.txt`

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

<img width="425" alt="image" src="https://user-images.githubusercontent.com/70572044/217575589-dd909f69-f8a8-4350-98ad-a747ebf54248.png">

ndeloof committed:
Introduce support for docker commit

Signed-off-by: Laura Brehm <laurabrehm@hey.com>
Co-authored-by: Laura Brehm <laurabrehm@hey.com>
Co-authored-by: Sebastiaan van Stijn <github@gone.nl>
Co-authored-by: Paweł Gronowski <pawel.gronowski@docker.com>
Co-authored-by: Nicolas De Loof <nicolas.deloof@gmail.com>

laurazard commented on daemon/containerd/image_commit.go:
```diff
@@ -1,17 +1,296 @@
+
+	cs := i.client.ContentStore()
+
+	ocimanifest, err := images.Manifest(ctx, cs, desc, platforms.DefaultStrict())
```
@thaJeztah In the upstream we were using `ContainerConfig.Platform` https://github.com/rumpl/moby/blob/2670c94d8298523ca90db1b097d0c4ebfc3ccc56/daemon/containerd/image_commit.go#L50, but this didn't get upstreamed in https://github.com/moby/moby/pull/44804 – related comment here https://github.com/rumpl/moby/pull/98#discussion_r990187381

thaJeztah replied:
Hmm.. right, would have to try to think what makes sense here; let me at least start with a quick summary of my thoughts (in case it helps);

- `docker commit` / `docker container commit` commits a _container's_ filesystem to an image
- the container would always be running a _single_ arch, so the arch should match the arch that the container was started for
- are there any cases that we can think of where the code here would be different from `ContainerConfig.Platform` ?

>  related comment here https://github.com/rumpl/moby/pull/98#discussion_r990187381

Hm, right, so that comment was about distinction between `Config` and `HostConfig`. Generally `HostConfig` would hold the "non portable" parts of the configuration; see https://github.com/moby/moby/blob/3ba527d82af53cf752aa6ec39daac14626c41b0f/api/types/container/hostconfig.go#L375-L378

Admitted, the line between "portable" and "non-portable" is sometimes "blurry" (maybe we need a better description/definition there), but I think in this case `Platform` would be considered "non-portable" (it doesn't _immediately_ define what the container looks like; I'm able to run the same container on a different architecture if the image is multi-arch, and the corresponding architecture can be resolved). (/cc @tianon @cpuguy83 for thoughts).

vvoland replied:
I just noticed I had a pending comment in rumpl#98 >.<
I agree that having `Platform` in HostConfig would make more sense. It seems that HostConfig is not serialized though:
https://github.com/moby/moby/blob/2670c94d8298523ca90db1b097d0c4ebfc3ccc56/container/container.go#L88

And we need to persist the platform because we need it for restoring containers after daemon restart.
Possibly this should be in `Container` struct itself, if we can't have it in HostConfig?

cpuguy83 replied:
I don't think there's any reason to think the commited fs would have a different platform.
Also all these platform fields are codified in the OCI image config spec anyway.
If we really need it (like down the line) it could be optionally passed down from the client.

vvoland commented on daemon/containerd/image_commit.go:
```diff
@@ -1,17 +1,296 @@
+	}
+
+	configDesc := ocispec.Descriptor{
+		MediaType: images.MediaTypeDockerSchema2Config,
```
+1 . We already use the OCI media types for `docker import`.
https://github.com/moby/moby/blob/3ba527d82af53cf752aa6ec39daac14626c41b0f/daemon/containerd/image_import.go#L114

laurazard replied:
Will do :)

tianon replied:
+1 :smile:

laurazard committed:
Resolve and store manifest when creating container

This addresses the previous issue with the containerd store where, after a container is created, we can't deterministically resolve which image variant was used to run it (since we also don't store what platform the image was fetched for).

This is required for things like `docker commit`, and computing the containers layer size later, since we need to resolve the specific image variant.

Signed-off-by: Laura Brehm <laurabrehm@hey.com>

laurazard commented:
Just pushed a commit addressing the issues discussed in the maintainers call (and referenced in https://github.com/moby/moby/pull/44934#discussion_r1102022773).

It resolves the root issue of not being able to deterministically resolve the image variant used to run a container after the container is created by storing the digest of the image-variant digest, and then using that later when necessary to fetch the content from the store and using that instead of making assumptions about the platform.

I'd appreciate some input about the design itself - I tried avoiding introducing another `daemon.UsesSnapshotter()` and a containerd specific `ImageService` method, but in the end this seemed the cleanest way to (for now) achieve the desired outcome. Maybe in the future another abstraction will make more sense. The specific refactor code is in this commit – https://github.com/moby/moby/pull/44958/commits/867ec3db2cfd6936d2113823995f37338b500749. We can squash it later, but I thought it better to leave it seperate for now to make discussion clearer.

cc @neersighted @thaJeztah

vvoland commented on daemon/containerd/image.go:
```diff
@@ -122,6 +122,36 @@ func (i *ImageService) GetImage(ctx context.Context, refOrID string, options ima
+
+	platform := platforms.AllPlatformsWithPreference(cplatforms.Default())
+	if options.Platform != nil {
+		platform = cplatforms.OnlyStrict(*options.Platform)
```
This one should be `cplatforms.Only` to match `PrepareSnapshot` behavior. We want the "lower" variant images to be usable on "higher" variant platforms (example: `linux/arm/v5`  will run fine on `linux/arm/v6`).

tianon replied:
Hah, high five - same thoughts at the same time (https://github.com/moby/moby/pull/44958#discussion_r1123631350). :sunglasses:

vvoland replied:
Oh, indeed! I didn't notice it at the time I was posting my comment 😆

laurazard replied:
Ooh, missed that, thanks ❤️

thaJeztah commented on container/container.go:
```diff
@@ -61,22 +61,23 @@ type ExitStatus struct {
+	Args               []string
+	Config             *containertypes.Config
+	ImageID            image.ID `json:"Image"`
+	ImageVariantDigest string
```
Some discussion around the name (bike shed!); let's give people some time to comment alternatives 👍  (also a comment, and perhaps an explicit `json:...`)

neersighted replied:
As `ImageVariantDigest` isn't obvious in a vacuum/outside this diff, I'd prefer something like `ImageResolvedRef` myself.

tianon commented on daemon/containerd/image.go:
```diff
@@ -122,6 +122,36 @@ func (i *ImageService) GetImage(ctx context.Context, refOrID string, options ima
+
+	platform := platforms.AllPlatformsWithPreference(cplatforms.Default())
+	if options.Platform != nil {
+		platform = cplatforms.OnlyStrict(*options.Platform)
```
Is this function ever going to be used in a way that it _should_ respect the vector for the supplied platform? (`platforms.Only` vs `platforms.OnlyStrict`)

tianon commented on daemon/containerd/image.go:
```diff
@@ -122,6 +122,36 @@ func (i *ImageService) GetImage(ctx context.Context, refOrID string, options ima
 	return img, nil
 }
 
+func (i *ImageService) GetImageVariant(ctx context.Context, refOrID string, options imagetype.GetImageOpts) (*ocispec.Descriptor, error) {
```
As discussed a  little bit elsewhere in this PR, I think "variant" is a _slightly_ confusing choice for this function since that's used in OCI for the "v7" part of a platform value like `linux/arm/v7` (`{"os":"linux","architecture":"arm","variant":"v7"}`), but I don't think this is _too_ sticky of a point as long as we don't write it to disk anywhere or encode it in an API we can't change or drop later. :sweat_smile:

Maybe get `GetPlatformImage` or something like that to make it clear this is to take a potentially-index/list ref and return a specific descriptor for a "resolved" single platform image instead?

tianon replied:
I guess it could even be `GetImageManifest` (matching @thaJeztah's suggestion for the related field name) :eyes:

thaJeztah replied:
`GetImageManifest` or `ResolveImageManifest` works for me (naming is hard!)

neersighted replied:
I like `GetImageManifest` following the principle of symmetry with the 'legacy'/less specific counterpart (e.g. `Image`, `ImageManifest`).

laurazard replied:
Sounds great! :)

laurazard commented:
Github is having some issues with the review requests 👀

laurazard commented:
> Step 6/13 : RUN mkdir C:\tmp && mkdir C:\bin
> unexpected EOF
Looks unrelated

laurazard commented:
> === FAIL: amd64.integration.build TestBuildMultiStageParentConfig (0.01s)
>    build_test.go:220: assertion failed: error is not nil: unexpected EOF
Lots of `EOF` in the integration tests (in Windows and Ubuntu), but I'm not sure what could be causing it.

laurazard commented:
Oooooops, didn't realize @neersighted managed to request the reviews (as they hadn't updated on my end) and they disappeared again when I tried to do them 😨 Github why.

vvoland reviewed (approved):
LGTM; there's just one little debug log that's left 😄

laurazard commented on daemon/create.go:
```diff
@@ -70,6 +70,8 @@ func (daemon *Daemon) containerCreate(ctx context.Context, opts createOpts) (con
 		img, err := daemon.imageService.GetImage(ctx, opts.params.Config.Image, imagetypes.GetImageOpts{Platform: opts.params.Platform})
 		if err != nil {
+			println("=====================")
+			println("oh neeeeeeeeein")
```
oh noooooooo

thaJeztah commented:
code LGTM; given that the first 5 commits are mostly fixing up previous changes that were added in this PR, perhaps we should squash the first 5 ones (last one may be ok to keep separate, as it's additional functionality); WDYT @vvoland @laurazard ?

vvoland commented:
+1 on squashing

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
Hmm... looks like CI completed, but still shows as "running" on GitHub? https://github.com/moby/moby/actions/runs/4344357044/jobs/7587778593

<img width="518" alt="Screenshot 2023-03-06 at 16 18 28" src="https://user-images.githubusercontent.com/1804568/223152793-790d5815-57db-4c01-a982-4200f3d97383.png">
<img width="1434" alt="Screenshot 2023-03-06 at 16 19 48" src="https://user-images.githubusercontent.com/1804568/223152803-0de4a85c-db80-4a7c-bf78-0403c070ce72.png">

thaJeztah commented:
Oh! Maybe that's because https://github.com/moby/moby/pull/44934 also includes that commit, which confuses GHA?

laurazard commented:
Just Github things 😅
